### PR TITLE
ConfigObjectUtility::GetObjectConfigPath(): hash names of not already existing objects

### LINKED ISSUE #8022: Can't add a comment or schedule a downtime for a host/service with a long name
## Describe the bug

Adding a comment or a downtime for a host or service with a long name shows this API error:

```
icinga2: Can't send external Icinga command: 500 Action execution failed: 'Error: Function call 'std::ifstream::open' for file '/opt/icinga2/var/lib/icinga2/api/packages/_api/07d2d10f-dd07-4159-a60a-a0323aa6fe4d/conf.d/comments/looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong-hoooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooost!looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong-seeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeerrrrrrrrrrrrrrrviiiiiiiiiiiiiiiiice!d456a095-d6fb-467b-b196-88694f059537.conf' failed with error code 36, 'File name too long'

(0) Compiling configuration file '/opt/icinga2/var/lib/icinga2/api/packages/_api/07d2d10f-dd07-4159-a60a-a0323aa6fe4d/conf.d/comments/looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong-hoooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooost!looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong-seeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeerrrrrrrrrrrrrrrviiiiiiiiiiiiiiiiice!d456a095-d6fb-467b-b196-88694f059537.conf'
'.
```

## Expected

If Icinga can handle long host/services names in general, I'd expect it to handle them as well when interacting with the API.

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): r2.12.0-rc1-1
* Operating System and version: CentOS 7

### PR DESCRIPTION
**Pull request by Al2Klimov**
... to avoid too long file names.

fixes #8022

### CODE REVIEW

**Comment by Yonas-net** [comment]
Testprotocol:
So far Icinga2 has not been able to process api request error-free with a very long service or host name.

The host object that i use for this test looks like as folllows..
```
object Host "LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg" {
  address = "127.0.0.1"
  check_command = "http"
}
```
when I tried to add a comment to this host with api query this error appears...
Command: ``curl -k -v -u root:icinga -H 'Accept: application/json'  -X POST 'https://localhost:5665/v1/actions/add-comment' -d '{ "type": "Host", "filter": "host.name==\"LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg\"", "author": "yonas", "comment": "Some comments here"}'``
and the ``ERROR`` looks like below..
```
"Action execution failed: 'Error: Function call 'std::ifstream::open' for file '/home/yonas/Workspace/icinga2/prefix/var/lib/icinga2/api/packages/_api/41e8a3da-8ed0-45a5-a4bd-f5ed0dbcc531/conf.d/comments/LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!10c30cc8-df21-436c-8ab5-913ef23b104d.conf' failed with error code 36, 'File name too long'\n\n\t(0) Compiling configuration file '/home/yonas/Workspace/icinga2/prefix/var/lib/icinga2/api/packages/_api/41e8a3da-8ed0-45a5-a4bd-f5ed0dbcc531/conf.d/comments/Loooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn* Connection #0 to host localhost left intact
nnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!10c30cc8-df21-436c-8ab5-913ef23b104d.conf'\n'."
```
but now icinga2 does this correctly.
```
{"results":[{"code":200.0,"legacy_id":5.0,"name":"LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!de83db38-284b-4ba5-9ce1-9ba9b24f2a04","status":"Successfully added comment 'LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!de83db38-284b-4ba5-9ce1-9ba9b24f2a04* Connection #0 to host localhost left intact
' for object 'LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg'."}]}
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
I'm not sure about this as this completely removes the ability to know from the filename what's in the file. I'd rather prefer something like a truncated name + hash.

File: lib/remote/configobjectutility.cpp [hunk 1]
```diff
@@ -35,8 +37,13 @@ String ConfigObjectUtility::GetObjectConfigPath(const Type::Ptr& type, const Str
...
 
-	return prefix + "/conf.d/" + typeDir +
-		"/" + EscapeName(fullName) + ".conf";
+	auto old (prefix + "/conf.d/" + typeDir + "/" + EscapeName(fullName) + ".conf");
+
+	if (Utility::PathExists(old)) {
+		return std::move(old);
+	}
+
+	return prefix + "/conf.d/" + typeDir + "/" + fullName.SubStr(0, 10) + SHA1(fullName) + ".conf";
```
Comment (julianbrost) [inline lib/remote/configobjectutility.cpp:46]: 10 seems a bit short to me. Looking at https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits, I think we can safely assume a limit of 255 bytes. We don't have to max out that limit, but something around 80 should be enough to represent the full name for most setups. Also some separator between name and hash would be nice.

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Review by julianbrost — APPROVED** [review]
Fine for me, paths now look like this:

```
/var/lib/icinga2/api/packages/_api/77e78991-06d4-4589-aa0c-ee60e1d7724e/conf.d/downtimes/host-24590-ytwnaddqifgyxfjjrfrmijtlrqekhatcdeomrjlsrlyehcwcuhcndvaabbczdishqejhr...aed80d7856fd65df4d9ac22ad61cc37afb81b7b4.conf
/var/lib/icinga2/api/packages/_api/77e78991-06d4-4589-aa0c-ee60e1d7724e/conf.d/services/host-24590-ytwnaddqifgyxfjjrfrmijtlrqekhatcdeomrjlsrlyehcwcuhcndvaabbczdishqejhr...db08056520e94625048fe05e3ef6028d869a27eb.conf
/var/lib/icinga2/api/packages/_api/77e78991-06d4-4589-aa0c-ee60e1d7724e/conf.d/hosts/host-24590-ytwnaddqifgyxfjjrfrmijtlrqekhatcdeomrjlsrlyehcwcuhcndvaabbczdishqejhr...3d0cbea021b186417dcf1d494e456c951fda4ced.conf
```

@Yonas-net Please take another look at this since this changed after your approval.

**Review by Yonas-net — APPROVED** [review]
LGTM!
```
{"results":[{"code":200.0,"legacy_id":1.0,"name":"LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!fb57e44f-75a2-4af2-8dc1-d3b960f6868f","status":"Successfully added comment 'LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg!fb57e44f-75a2-4af2-8dc1-d3b960f6868f' for object 'LoooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggggggggggggggggggggggggggMoooooooooooooooooooooooooooooooooooooooooooooorrrrrrrrrrrrrrrrrrrrrrrrrrrreeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeLooooooooooooooooooooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggggggggggggggggg'."}]}
```

**Pull request merged**